File: delayedqueue-jvm/src/main/kotlin/org/funfix/delayedqueue/jvm/CronConfigHash.kt
package org.funfix.delayedqueue.jvm

import java.security.MessageDigest
import java.time.Duration

/**
 * Hash of a cron configuration, used to detect configuration changes.
 *
 * When a cron schedule is installed, this hash is used to identify messages belonging to that
 * configuration. If the configuration changes, the hash will differ, allowing the system to clean
 * up old scheduled messages.
 *
 * @property value the MD5 hash string
 */
@JvmRecord
public data class CronConfigHash(public val value: String) {
    override fun toString(): String = value

    public companion object {
        /** Creates a ConfigHash from a daily cron schedule configuration. */
        @JvmStatic
        public fun fromDailyCron(config: CronDailySchedule): CronConfigHash {
            val text = buildString {
                appendLine() // Leading newline to match Scala stripMargin
                appendLine("daily-cron:")
                appendLine("  zone: ${config.zoneId}")
                appendLine("  hours: ${config.hoursOfDay.joinToString(", ")}")
            }
            return CronConfigHash(md5(text))
        }

        /** Creates a ConfigHash from a periodic tick configuration. */
        @JvmStatic
        public fun fromPeriodicTick(period: Duration): CronConfigHash {
            val text = buildString {
                appendLine() // Leading newline to match Scala stripMargin
                appendLine("periodic-tick:")
                appendLine("  period-ms: ${period.toMillis()}")
            }
            return CronConfigHash(md5(text))
        }

        /** Creates a ConfigHash from an arbitrary string. */
        @JvmStatic public fun fromString(text: String): CronConfigHash = CronConfigHash(md5(text))

        private fun md5(input: String): String {
            val md = MessageDigest.getInstance("MD5")
            val digest = md.digest(input.toByteArray())
            return digest.joinToString("") { "%02x".format(it) }
        }
    }
}

File: delayedqueue-scala/jvm/src/main/scala/org/funfix/delayedqueue/scala/CronConfigHash.scala
/*
 * Copyright 2026 Alexandru Nedelcu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.funfix.delayedqueue.scala

import java.security.MessageDigest
import org.funfix.delayedqueue.jvm

/** Hash of a cron configuration, used to detect configuration changes.
  *
  * When a cron schedule is installed, this hash is used to identify messages belonging to that
  * configuration. If the configuration changes, the hash will differ, allowing the system to clean
  * up old scheduled messages.
  *
  * @param value
  *   the MD5 hash string
  */
final case class CronConfigHash(value: String) {
  override def toString: String = value

  /** Converts this Scala CronConfigHash to a JVM CronConfigHash. */
  def asJava: jvm.CronConfigHash =
    new jvm.CronConfigHash(value)
}

object CronConfigHash {

  /** Creates a ConfigHash from a daily cron schedule configuration. */
  def fromDailyCron(config: CronDailySchedule): CronConfigHash = {
    // Port from Kotlin: buildString { appendLine(); appendLine("daily-cron:"); ... }
    val text = new StringBuilder()
    text.append('\n') // Leading newline to match Kotlin
    text.append("daily-cron:\n")
    text.append(s"  zone: ${config.zoneId}\n")
    text.append(s"  hours: ${config.hoursOfDay.mkString(", ")}\n")
    CronConfigHash(md5(text.toString))
  }

  /** Creates a ConfigHash from a periodic tick configuration. */
  def fromPeriodicTick(period: java.time.Duration): CronConfigHash = {
    // Port from Kotlin: buildString { appendLine(); appendLine("periodic-tick:"); ... }
    val text = new StringBuilder()
    text.append('\n') // Leading newline to match Kotlin
    text.append("periodic-tick:\n")
    text.append(s"  period-ms: ${period.toMillis}\n")
    CronConfigHash(md5(text.toString))
  }

  /** Creates a ConfigHash from an arbitrary string. */
  def fromString(text: String): CronConfigHash = CronConfigHash(md5(text))

  private def md5(input: String): String = {
    val md = MessageDigest.getInstance("MD5")
    val digest = md.digest(input.getBytes)
    digest.map("%02x".format(_)).mkString
  }

  /** Conversion extension for JVM CronConfigHash. */
  extension (javaHash: jvm.CronConfigHash) {

    /** Converts a JVM CronConfigHash to a Scala CronConfigHash. */
    def asScala: CronConfigHash =
      CronConfigHash(javaHash.value)
  }
}
